### DISCUSSION TITLE
Including blank spaces in custom variables

### DISCUSSION THREAD

**Discussion by profucius** [Taitava/obsidian-shellcommands#317] [Q&A] [answered]
I tried looking through the discussions and documentation for this:  Is it possible to include a blank space when setting a custom variable?

So far I have had to work around this by using an underscore instead, but it gets complicated because I am trying to reference files/folders with the variable, so I have to use my macro app to replace the underscore, to and from, each time.

I'll laugh if its something simple! I did try using double and single quotes, but it seems to ignore those.

Just for clarity sake, this is an example of what I try:  `obsidian://shell-commands/?vault=Vault&_customvariable=Two Words`
The result I get is `Two`
And if I try: `obsidian://shell-commands/?vault=Vault&_customvariable="Two Words"`
The result I get is `"Two`

**Comment by Taitava** [comment, marked as answer]
Hi,

you need to URL encode all special characters in the value (and space is considered as a special character here). For this particular use-case, you can just replace space with `%20`, so your url would become `obsidian://shell-commands/?vault=Vault&_customvariable=Two%20Words`.

To be further on the safe side, you can use some online URL encoder. It's good in case your values happen to contain also other special characters than just spaces. This one seems to be a simple and easy to use URL encoder: https://www.urlencoder.io/ . Just write the **value** into the text field and it will give you the url encoded version of the value immediately. Note that you should only write the value to the encoder - **not** other parts of the url, because that would break the important `/`, `?`, `=` and `&` characters in the URL.

- I need to add this information to the [documentation page of Shell commands URI](https://publish.obsidian.md/shellcommands/Basic+usage/Shell+commands+URI). I just realised it does not tell anything about URL encoding, which is quite a crucial aspect of the feature. (Edit: Done).
- I also need to think about some kind of feature for SC that would ask values from a user and craft URLs that have the values encoded automatically.

I'll move this discussion to _Q & A_. 👍

**Reply by profucius** [reply]
Thanks for the reply. I just tried this, and it does not seem to work unfortunately. I followed your instructions exactly, and within the SC custom variables pane I can see the result is `Two Words` however I then run a powershell script to store that into an environmental variable (I'll post the exact script line below) but the result I get from it is ``Two` Words`` with a backtick and space between them.

Powershell script (ran as a SC): `[System.Environment]::SetEnvironmentVariable('EnvVariable','{{_customvariable}}', 'Machine')`

This script worked fine when I was using an underscore between the two words, but using %20 instead (and having generated it from the urlencoder link you shared) injects that backtick.

Could I have done something wrong in my testing?

Update: I tried using CMD instead of Powershell using the below command, and it worked properly. So I wonder if something is going wrong in passing the data via Powershell 5/Core. In the meantime I'll start using the CMD method as it doesn't cause trouble for my workflow (now that I'm aware of using & to do multiple at once).

`SETX "EnvVariable" "{{_customvariable}}" -m`

**Reply by Taitava** [reply]
It looks like we got a bit further, solved one issue, but ran into another. I hope my answer here solves the new problem, too. 🙂 

The backtick `` ` `` comes from the SC plugin when using PowerShell as a shell. `{{variables}}` place escape characters (`` ` `` for PowerShell) in front of all characters that are not `a`-`z`, `A`-`Z`, `0`-`9`, nor `_`. It seems that PowerShell does not remove the escape characters when the variable value is wrapped in single quotes `'`. I tried to remove the single quotes completely, but got a PowerShell error message saying `Missing expression after ','` but I got it working when I changed the single quotes to double quotes `"`:
```PowerShell
[System.Environment]::SetEnvironmentVariable('EnvVariable',"{{_customvariable}}", 'Machine')
```
(For me, I also got an error message `Exception calling "SetEnvironmentVariable" with "3" argument(s): "Requested registry access is not allowed."` when I had the `'Machine'` argument in place, but I removed the argument so it worked for me. But I guess the argument works ok for you?)

The backticks are there for the idea that special characters should be treated non-specially, as literal characters. E.g. a single/double quote character in `{{_customvariable}}` should not cause the closing of an opened single/double quote in the command. The reason your CMD version worked fine is because I haven't found a working solution for doing escaping in CMD, so the `SC` plugin does not add any escaping characters when the shell is CMD. In case escaping in PowerShell causes any problems you can't circumvent, you can always start variables with `{{!` instead of `{{` to prevent escaping. But I recommend trying to find other solutions first, as non-escaped variable values can lead to accidental bugs in your commands. [More about escaping](https://publish.obsidian.md/shellcommands/Variables/Escaping+special+characters+in+variable+values).
